File: 基于ESP32的智能小车控制/CarControl.cpp
#include <RL_ESP32_Motor.h>
#include <ESP32_Servo.h>
#include <HardwareSerial.h>
#include <ArduinoJson.h>


// 定义电机引脚
Motor Esp32_Motor_1(1,12,13);
Motor Esp32_Motor_2(2,14,15);
Motor Esp32_Motor_3(3,16,17);
Motor Esp32_Motor_4(4,18,19);

// 创建舵机实例
Servo servo_25;
Servo servo_26;

//定义超声波测距引脚
const int trigPin = 4;
const int echoPin = 5;

float distance;
int speed;

// 小车前进
void advance(int speed = speed) {
  Esp32_Motor_1.Motor_Speed(speed);
  Esp32_Motor_2.Motor_Speed(speed);
  Esp32_Motor_3.Motor_Speed((-speed));
  Esp32_Motor_4.Motor_Speed((-speed));
}


// 小车后退
void back(int speed = speed) {
  Esp32_Motor_1.Motor_Speed((-speed));
  Esp32_Motor_2.Motor_Speed((-speed));
  Esp32_Motor_3.Motor_Speed(speed);
  Esp32_Motor_4.Motor_Speed(speed);
}

// 小车停止
void stop() {
  Esp32_Motor_1.Motor_Speed(0);
  Esp32_Motor_2.Motor_Speed(0);
  Esp32_Motor_3.Motor_Speed(0);
  Esp32_Motor_4.Motor_Speed(0);
}


// 小车左转
void left(int speed = speed, float degree = 0.25) {
  Esp32_Motor_1.Motor_Speed(speed);
  Esp32_Motor_2.Motor_Speed(speed);
  Esp32_Motor_3.Motor_Speed((-speed * (1 + degree)));
  Esp32_Motor_4.Motor_Speed((-speed * (1 + degree)));
}

// 小车右转
void right(int speed = speed, float degree = 0.25) {
  Esp32_Motor_1.Motor_Speed(speed * (1 + degree));
  Esp32_Motor_2.Motor_Speed(speed * (1 + degree));
  Esp32_Motor_3.Motor_Speed((-speed));
  Esp32_Motor_4.Motor_Speed((-speed));
}

// 逆时针旋转
void anticlockwise(int speed = speed) {
  Esp32_Motor_1.Motor_Speed((-speed));
  Esp32_Motor_2.Motor_Speed((-speed));
  Esp32_Motor_3.Motor_Speed((-speed));
  Esp32_Motor_4.Motor_Speed((-speed));
}

// 顺时针旋转
void clockwise(int speed = speed) {
  Esp32_Motor_1.Motor_Speed((speed));
  Esp32_Motor_2.Motor_Speed((speed));
  Esp32_Motor_3.Motor_Speed((speed));
  Esp32_Motor_4.Motor_Speed((speed));
}

// 向左平移
void left_tran(int speed = speed) {
  Esp32_Motor_1.Motor_Speed((-speed));
  Esp32_Motor_2.Motor_Speed((speed));
  Esp32_Motor_3.Motor_Speed((speed));
  Esp32_Motor_4.Motor_Speed((-speed));
}

// 斜向左前方
void left_oblique(int speed = speed) {
  Esp32_Motor_1.Motor_Speed(0);
  Esp32_Motor_2.Motor_Speed((speed));
  Esp32_Motor_3.Motor_Speed(0);
  Esp32_Motor_4.Motor_Speed((-speed));
}

// 向右平移
void right_tran(int speed = speed) {
  Esp32_Motor_1.Motor_Speed((speed));
  Esp32_Motor_2.Motor_Speed((-speed));
  Esp32_Motor_3.Motor_Speed((-speed));
  Esp32_Motor_4.Motor_Speed((speed));
}

// 斜向右前方
void right_oblique(int speed = speed) {
  Esp32_Motor_1.Motor_Speed((speed));
  Esp32_Motor_2.Motor_Speed(0);
  Esp32_Motor_3.Motor_Speed((-speed));
  Esp32_Motor_4.Motor_Speed(0);
}

// 舵机转动
void servoset(int d1 = 90, int d2 = 100){
  servo_25.write(d1);
  delay(100);
  servo_26.write(d2);
}

// 超声波测距
float checkDistance() {
  digitalWrite(trigPin, LOW);
  delayMicroseconds(2);
  digitalWrite(trigPin, HIGH);
  delayMicroseconds(10);
  digitalWrite(trigPin, LOW);

  float distance = pulseIn(echoPin, HIGH) / 58.00;   //echo time conversion into a distance
  delay(10);
  return distance;
}

String messageIn;
HardwareSerial mySerial1(1);

void setup(){

  mySerial1.begin(115200,SERIAL_8N1,32,33); // 自定义32，33引脚为串口
  speed = 32;
  Esp32_Motor_1.mcpwm_begin();
  Esp32_Motor_2.mcpwm_begin();
  Esp32_Motor_3.mcpwm_begin();
  Esp32_Motor_4.mcpwm_begin();

  servo_25.attach(25,500,2500);
  servo_26.attach(26,500,2500);

  pinMode(trigPin, OUTPUT);
  pinMode(echoPin, INPUT);
  distance = 0;
}

void loop(){
    //声明c串口通信字符串变量
    String rx_buffer = "";
    while (mySerial1.available() > 0) //串口接收到数据
    {
        rx_buffer += char(mySerial1.read());
        delay(2);
    }

    if (rx_buffer.length() > 0)
    {
        DynamicJsonDocument  jsonBuffer(400);
        deserializeJson(jsonBuffer, rx_buffer); // 解析Json数据
        JsonObject doc = jsonBuffer.as<JsonObject>();

        int speed_0 = doc["speed"][0]; // "0/1/2"
        int speed_1 = doc["speed"][1];
        const char* carmove = doc["carmove"][0]; // "STOP/ADVANCE/LEFT/..."
        float degree = doc["carmove"][1];
        int servo_0 = doc["servo"][0];
        int servo_1 = doc["servo"][1];
        int checkdistance = doc["checkdistance"]; // "1/0"

        // 执行速度调节
        if (speed_0 == 0 and 0 <= speed_1 <= 100)
            speed = speed_1;
        else if(speed_0 == 1)
            speed = speed + 10;
        else if (speed_0 == -1)
            speed = speed - 10;

        // 执行运动控制
        messageIn = carmove;
        if (messageIn == "ADVANCE")
            advance(speed);
        else if (messageIn == "BACK")
            back(speed);
        else if (messageIn == "STOP")
            stop();
        else if (messageIn == "LEFT" and 0 < degree <= 2)
            left(speed, degree);
        else if (messageIn == "RIGHT" and 0 < degree <= 2)
            right(speed, degree);
        else if (messageIn == "LEFT_TRAN")
            left_tran(speed);
        else if (messageIn == "RIGHT_TRAN")
            right_tran(speed);
        else if (messageIn == "ANTICLOCKWISE")
            anticlockwise(speed);
        else if (messageIn == "CLOCKWISE")
            clockwise(speed);
        else if (messageIn == "LEFT_OBLIQUE")
            left_oblique(speed);
        else if (messageIn == "RIGHT_OBLIQUE")
            right_oblique(speed);

        // 执行舵机控制
        servoset(servo_0, servo_1);

        // 执行超声波测距
        if (checkdistance == 1){
            distance = checkDistance();
            // 返回Json数据
            DynamicJsonDocument data(256);
            data["distance"] = distance;
            char json_string[256];
            serializeJson(data, json_string); // 打包成Json串
            mySerial1.println(json_string); // 串口输出
        }

    }

}

File: 基于ESP32的智能小车控制/carControl.py
# -*- coding: utf-8 -*-
"""
carControl.py使用说明：
========
import carControl
1.速度设置 speed(value: 'str|int')
    设置速度为 0~100的任意数值：carControl.speed(50)
    加速（v+10）：carControl.speed('up')
    减速(v-10)：carControl.speed('down')
    注意：小车电机有起始转速，不同场地因摩擦力不同最低启动速度也不同，第一版道路场景中 直行的启动速度为>=32，左平移则>=46

2.方向控制 move(value, degree=0)
    moveList = ['ADVANCE', #前进
                'BACK', #后退
                'STOP', #停车
                'LEFT_TRAN', #左平移
                'RIGHT_TRAN', #右平移
                'CLOCKWISE', #顺时针旋转
                'ANTICLOCKWISE', #逆时针旋转
                'LEFT_OBLIQUE', #左斜45°前移
                'RIGHT_OBLIQUE'] #右斜45°前移
    简单的方向控制：carControl.move(value in moveList)
    左右转向需加上转角幅度 degree：carControl.move('LEFT'/'RIGHT', 0.5)
    degree代表转向幅度，取值0~2，数值与转向幅度成正比

3.舵机控制 servo(d1, d2)
    d1、d2为两个舵机转角，0~180: carControl.servo(90, 90)
    d1为相机垂直方向的角度，d1=90°约为水平；d2为相机水平方向的角度，d2=90°为正前方
    小车摄像头平视前方为[90，90]，道路循迹行驶合适的视角为[110, 90]

4、超声波测距 check_distance()，将返回超声波模块正前方第一个障碍物的距离值，可测范围5~500，单位cm
    carControl.check_distance()

5、小车执行连续的两个命令的时间间隔约为0.08s，可连续发送指令，无需再设置等待时间.
"""
import json
import serial

# 初始化默认串口信息
speedInfo = {'speed': [-100, -100]}
moveInfo = {'carmove': ['RELEX', 0]}
servoInfo = {'servo': [-1, -1]}
checkDistanceInfo = {'checkdistance': 0}

ser = serial.Serial('/dev/ttyAMA1', 115200)  # A200 DK与小车间的接口定义
# ser = serial.Serial('COM5',115200)  # windos串口设置


# 串口发送Json数据
def send_data_to_serial():
    # 将所有数据打包成Json格式
    data = dict(
        list(speedInfo.items()) + list(moveInfo.items()) + list(servoInfo.items()) + list(checkDistanceInfo.items()))
    json_data = json.dumps(data)
    ser.write(json_data.encode())  # 将Json格式的数据发送至串口


def speed(value: 'str|int'):
    if value == 'up':
        speedInfo['speed'] = [1, 10]
    elif value == 'down':
        speedInfo['speed'] = [-1, 10]
    else:
        speedInfo['speed'] = [0, value]
    send_data_to_serial()
    return read_serial()


def move(value, degree=0):
    moveList = ['ADVANCE', 'BACK', 'STOP', 'LEFT_TRAN', 'RIGHT_TRAN', 'CLOCKWISE', 'ANTICLOCKWISE', 'LEFT_OBLIQUE',
                'RIGHT_OBLIQUE']
    if value in moveList:
        moveInfo['carmove'] = [value, 0]
    elif value in ['LEFT', 'RIGHT'] and 0 < degree <= 2:
        moveInfo['carmove'] = [value, degree]
    send_data_to_serial()
    return read_serial()


def servo(d1, d2):
    if 0 <= d1 <= 180 and 0 <= d2 <= 180:
        servoInfo['servo'] = [d1, d2]
    send_data_to_serial()
    return read_serial()


def check_distance():
    checkDistanceInfo['checkdistance'] = 1
    send_data_to_serial()
    # 获取串口返回的distance值
    distance = read_serial("distance")
    return distance


# 读取串口返回信息
def read_serial(kind="speed"):
    data = json.loads(ser.readline())
    if kind == "distance":
        return data["distance"]
    else:
        return data["speed"]  # 默认返回speed
